### PR TITLE
Performance tests improvements <master> [7971]

### PR DESCRIPTION
**Pull request by EduPonz**
Reopens #1070 

Fix multiple subscribers in LatencyTest.
Add support for multiple subscribers in ThroughputTest.
Large data support.
Demands file for latency test.
SHM transport support.
Adds intraprocess for latency and throughput
Fix intraprocess throughput

### CODE REVIEW

**Comment by EduPonz** [comment]
I've asked @rsanchez15 to also look into this so we have a fresh pair of eyes

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2253/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2060/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2252/)

**Comment by i-and** [comment]
Hi! I didn't see any bug fixes for #1237 here. Did I miss something? I will specify specific code fragments that contain the issue:

In file ThroughputPublisher.cpp line 570:
throughput_type_ = new ThroughputType( **(uint16_t)** msg_size);

In file ThroughputSubscriber.cpp line 459:
throughput_type_ = new ThroughputType( **(uint16_t)** data_size_);

In file ThroughputTypes.hpp line 88:
typedef struct ThroughputType {
    uint32_t seqnum;
    std::vector<uint8_t> data;
    ThroughputType(
            **uint16_t**  number)
        : seqnum(0)
        , data(number,0)
    {
    }
} ThroughputType;

In file ThroughputTypes.cpp line 36:
payload->length = 8+ **(uint16_t)** lt->data.size();

Why this type casting to uint16_t? If the msg_size or data_size_ or lt->data.size() is larger than 0xFFFF bytes, this cause the required sample size to be truncated.

I suggest you fix it by excluding the type cast to (uint16_t) and change the ThroughputType constructor argument type from uint16_t to uint32_t.

**Comment by EduPonz** [comment]
Hi @i-and ,

You were indeed totally right. Actually after checking #1070 , we did not have the proper fix there either. I've added it now so you should be able to run tests with sizes up to 2^32 bytes, I think that should be plenty for everyone.

I think there wasn't really a reason for the `uint16_t` other than during the design fragmentation was not a test-case.

Thanks anyway for your help

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2274/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2075/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2267/)

**Comment by MiguelCompany** [comment]
@EduPonz Please solve warning on Windows

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2311/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2111/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2303/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2344/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2142/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2335/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2346/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2144/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2337/)

**Comment by EduPonz** [comment]
This PR only modified performance and UDPv4 tests, so I'd any test failure of any other kind is unrelated.

File: test/performance/latency/LatencyTestPublisher.cpp [hunk 1]
```diff
@@ -333,11 +268,8 @@ bool LatencyTestPublisher::init(
     }
     subscriber_data_attributes.properties = property_policy;
 
-    if (large_data)
-    {
-        subscriber_data_attributes.historyMemoryPolicy =
-                eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
-    }
+    subscriber_data_attributes.historyMemoryPolicy =
+            eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestPublisher.cpp:272]: Does this fit into a single line?

File: test/performance/latency/LatencyTestSubscriber.cpp [hunk 2]
```diff
@@ -230,11 +212,8 @@ bool LatencyTestSubscriber::init(
     }
     subscriber_data_attributes.properties = property_policy;
 
-    if (large_data)
-    {
-        subscriber_data_attributes.historyMemoryPolicy =
-                eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
-    }
+    subscriber_data_attributes.historyMemoryPolicy =
+            eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
```
Comment (MiguelCompany) [inline test/performance/latency/LatencyTestSubscriber.cpp:216]: Does this fit into a single line?

File: test/performance/throughput/ThroughputPublisher.cpp [hunk 3]
```diff
@@ -61,14 +61,21 @@ void ThroughputPublisher::DataPubListener::onPublicationMatched(
...
         --throughput_publisher_.data_discovery_count_;
     }
 
     lock.unlock();
-    throughput_publisher_.data_discovery_cv_.notify_one();
+
+    if (throughput_publisher_.data_discovery_count_ == static_cast<int>(throughput_publisher_.subscribers_))
+    {
+        throughput_publisher_.data_discovery_cv_.notify_one();
+    }
```
Comment (MiguelCompany) [inline test/performance/throughput/ThroughputPublisher.cpp:78]: Should this be done before unlocking, as it is done below?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2364/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2164/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2355/)

**Comment by MiguelCompany** [comment]
Unrelated failures

**Pull request merged**